### ISSUE TITLE
ChallengeResponse null on server side

### ISSUE THREAD

**Issue by rbrecheisen** [restlet/restlet-framework-java#1040]
Hi,

We're trying to use the ChallengeResponse to add an OAuth2 access token to the HTTP request header. However, on the server-side the challenge response is always null. Now, we did find an old post (http://osdir.com/ml/java-restlet/2009-07/msg00132.html) where this issue was also reported and the person who posted it also proposed a work-around as follows:

```java
Series<Parameter> headers = new Form();
headers.add("Authorization", "Bearer " + accessToken);
clientProxy.getRequest().addAttributes().put("org.restlet.http.headers", headers);
```

After making the REST call to the server, I try to extract the access token in the server resource as follows:

```java
Series<Parameter> headers = getRequest().getAttributes().get("org.restlet.http.headers");
```

However, there's no entry "Authorization: Bearer" anymore. I'm wondering where did it go? Even so, this is a work around proposed some time ago (july 2009) but apparently the problem still persists in the latest version of Restlet (we're using 3.0-M1). We would much prefer using the "official" challenge response for this purpose.

Any ideas how we could get this working? Perhaps we should add more configuration options to the request? Would be great to get some feedback on this.

Kind regards,

Ralph

**Comment by rcbeuker** [comment]
some extra explanation;

We're trying to send Google Authentication tokens to our Rest interface on Google App Engine. 
As explained here (under heading "Calling a Google API")

https://developers.google.com/accounts/docs/OAuth2InstalledApp

We should do this by including an Authorization: Bearer HTTP header.
On the client side we're using;
```
    public ClientResource getClientResource(Reference pReference, String pAccessToken)
    {
        ClientResource lClientResource = new ClientResource(pReference);

        ChallengeResponse lChallengeResponse = new ChallengeResponse(ChallengeScheme.HTTP_OAUTH_BEARER);
        lChallengeResponse.setRawValue(pAccessToken);
        lClientResource.setChallengeResponse(lChallengeResponse);

        return lClientResource;
    }
```

So the ChallengeResponse with setRawValue() is used to set the header;
```
ChallengeResponse lChallengeResponse = new ChallengeResponse(ChallengeScheme.HTTP_OAUTH_BEARER);
lChallengeResponse.setRawValue(pAccessToken);
lClientResource.setChallengeResponse(lChallengeResponse);
```
For our J2SE clients this is working. Although there's an error message;

*"Challenge scheme HTTP_Bearer not supported by the Restlet engine."*

we're able to read the token on the server side with;
```
ChallengeResponse lChallengeResponse = pRequest.getChallengeResponse();
String lAccessToken = lChallengeResponse.getRawValue();
```
While this is working for our J2SE clients, our GWT client return always **lChallengeResponse == null**

Is this the correct method to set the *Authorization: Bearer HTTP header* or should we add extra/other configuration options? 

Why isn't this working for our GWT clients (return null)?

How do we get rid of the error message on our J2Se clients?

**Comment by rcbeuker** [comment]
I see a fix for HTTP_BASIC, is it possible to use this for our AccessToken?

http://restlet.tigris.org/ds/viewMessage.do?dsForumId=4447&dsMessageId=2368703

**Comment by rcbeuker** [comment]
The post (ChallengeResponse is NULL on the server side using GWT ) is from 2011, is this still an issue?

http://restlet.tigris.org/ds/viewMessage.do?dsForumId=4447&dsMessageId=2892504

**Comment by thboileau** [comment]
I've just tested locally (not on GAE), I see the challengeResponse on server side, and can display the raw value.
I try with GAE.

**Comment by thboileau** [comment]
As you said, I guess the fix (https://github.com/restlet/restlet-framework-java/commit/3b5b49e5739f7aff5c609a52eb6cfe4e2d70a95a) for this post explains what happens.

**Comment by thboileau** [comment]
I have a fix. I check with @jlouvel:
```java
            // If a challenge response is provided, update the GWT request builder
            ChallengeResponse challenge = request.getChallengeResponse();
            if (challenge != null) {
                if (ChallengeScheme.HTTP_BASIC.equals(challenge.getScheme())) {
                    // Handle manually the generation of the "Authorization"
                    // header, since the browser does not cope with that.
                    byte[] bytes = (challenge.getIdentifier() + ":" + String.valueOf(challenge.getSecret())).getBytes(CharacterSet.ISO_8859_1.getName());
                    result.getRequestBuilder().setHeader(
                            HeaderConstants.HEADER_AUTHORIZATION,
                            "Basic " + Base64.encode(bytes, false));
                } else if (challenge.getRawValue() != null && challenge.getScheme() != null) {
                    result.getRequestBuilder().setHeader(
                            HeaderConstants.HEADER_AUTHORIZATION,
                            challenge.getScheme().getTechnicalName() + " " + challenge.getRawValue());
                } else {
                    // In this case the login and password are simply
                    // concatenated to the URL.
                    result.getRequestBuilder().setUser(challenge.getIdentifier());
                    result.getRequestBuilder().setPassword(String.valueOf(challenge.getSecret()));
                }
            }

```

**Comment by jlouvel** [comment]
@thboileau The fix looks good to me.

**Comment by rbrecheisen** [comment]
Dear Jerome and Thierry,

How does this work the other way around? We're seeing a null challenge
response on the server side, not the GWT client side.

Regards Ralph

On Wednesday, March 25, 2015, Jerome Louvel <notifications@github.com>
wrote:

> @thboileau <https://github.com/thboileau> The fix looks good to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/restlet/restlet-framework-java/issues/1040#issuecomment-85804775>
> .
>


-- 
Sent from my iPhone

**Comment by thboileau** [comment]
Hi @rbrecheisen, that's the issue: the client does not send the Authorization header. I've tested this with an app deployed on GAE, and it works for me.

**Comment by rbrecheisen** [comment]
Hi Thierry,

Thanks for you quick reply. You're right of course, that _is_ the issue. I
was a bit confused. Still, I'm wondering. There's seems to be a challenge
response coming from the server which the client checks for null-ness. If
not null, the authorization header is set, as you indicated. However, in
our case, the GWT client authenticates itself directly with Google using
the gwt-oauth2 library (without going through the GAE server first). It
receives back an access token which we want to pass to the server with
every REST call, for example, to retrieve some data. The server can then
use the token to figure out who was calling (e.g., through the People API).
We could pass the access token as a REST parameter but we've heard that
this is not really safe. It's better to store the access token in the
Authorization Bearer header. So, how would this work in your case? We don't
have a challenge response object on the client side but just want to set
the above-mentioned header information before making the REST call. We
initially did that by instantiating a ChallengeResponse object in the
client ourselves (without getting one from the server) and calling
setRawValue(accessToken). However, in that case, the ChallengeResponse
received at the GAE server is null. If we set the header using
request.getAttributes().put("org.restlet.http.headers", headers) then the
"Authorization" parts gets thrown out and also never arrives at the server.

Maybe we're missing something really obvious here....

Kind regards,

Ralph

On Wed, Mar 25, 2015 at 9:19 AM, Thierry Boileau <notifications@github.com>
wrote:

> Hi @rbrecheisen <https://github.com/rbrecheisen>, that's the issue: the
> client does not send the Authorization header. I've tested this with an app
> deployed on GAE, and it works for me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/restlet/restlet-framework-java/issues/1040#issuecomment-85913387>
> .
>



-- 
Dr. Ir. Ralph Brecheisen

Brain Innovation B.V. + University of Maastricht
Oxfordlaan 55, 6229 EV
Maastricht (Netherlands)
*E* *brecheisen@brainvoyager.com <brecheisen@brainvoyager.com>*
*T* +31 6 46 230 127
http://nl.linkedin.com/in/ralphbrecheisen

**Comment by thboileau** [comment]
Hi Ralph,
thanks for coming back and sharing us your thoughts. I hope I'll have time this week to answer you.

**Comment by rcbeuker** [comment]
Hello Thierry,

Any news on this? We don't need a complete authorization workflow, we just want to transfer a Google Access Token to our GAE application via a HTTP header. This is working from a test J2SE client but not from GWT. 

Kind regards,

Roland

**Comment by thboileau** [comment]
Hello Roland,

I feel really sorry, it was difficult for me to spend some time on the framework.

>There's seems to be a challenge response coming from the server which the client checks for null-ness. If not null, the authorization header is set, as you indicated.

The framework has this semantic instead: 
- clients send ChallengeResponse to servers in order to authenticate the request.
- servers send ChallengeRequest to clients (in order to say to the client that its request misses right credentials), 

> It receives back an access token which we want to pass to the server with every REST call, for example, to retrieve some data. The server can then use the token to figure out who was calling (e.g., through the People API). We could pass the access token as a REST parameter but we've heard that
this is not really safe. It's better to store the access token in the Authorization Bearer header. So, how would this work in your case?

If I understand well, there is a first round trip, where the client ask for token, then this token is sent in every request the final server. In order to achieve this, the client need to instantiate a ChallengeResponse object with the token, with the "Bearer" scheme:

```java
ChallengeResponse lChallengeResponse = new ChallengeResponse(ChallengeScheme.HTTP_OAUTH_BEARER);
lChallengeResponse.setRawValue(pAccessToken);
contactResource.getClientResource().setChallengeResponse(lChallengeResponse);
```

and that's all.
Without the fix I mentioned above, such ChallengeResponse were not emittef by the GWT client. 
You should then be able to retrieve that challengeResponse on server side and read the access token.

Best regards,
Thierry Boileau

**Comment by rcbeuker** [comment]
Hello Thierry,

Thanks for your time! Our fault we are describing different problems pell-mell...

We are using the Bearer" scheme method as described above from our J2SE clients and its working. There's only a error message. I'll try your AuthenticatorHelper solution for this.

The real problem are the GWT clients. They're not transmitting the token information. Notable, the first round trip is between Google and our GWT clients without the use of Restlet. After this first round trip we have an acces token we want to transfer to GAE.

So we have an acces token, and we want to sent it to GAE like this;
```
public ClientResource getClientResource(Reference pReference, String pAccessToken)
    {
        ClientResource lClientResource = new ClientResource(pReference);

        ChallengeResponse lChallengeResponse = new ChallengeResponse(ChallengeScheme.HTTP_OAUTH_BEARER);
        lChallengeResponse.setRawValue(pAccessToken);
        lClientResource.setChallengeResponse(lChallengeResponse);

        return lClientResource;
    }
```

This is not working from GWT... Is there a Restlet patch for this?  Or is this fixed in the next Restlet versions?

Kind regards,

Roland Beuker

**Comment by thboileau** [comment]
Hello @rcbeuker,

this issue has been fixed in branch 2.3 by this commit https://github.com/restlet/restlet-framework-java/commit/20ebcd67849eb51ca24e1f31d3f6af77adab54e6. I will release the 2.3.2 this friday.
you can test it with the current 2.3 branch.

best regards,
Thierry Boileau

**Comment by thboileau** [comment]
Hi,
@rcbeuker , did you have time to test RF 2.3.2? Does it fix your issue?

I close this issue, since the fix should be fine, but feel free to reopen it!

Best regards,
Thierry Boileau
